Signal/src/util/OWSScreenLock.swift

```diff
@@ -0,0 +1,334 @@
+                                                                        comment: "Description of how and why Signal iOS uses Touch ID/Face ID to unlock 'screen lock'."),
+                                       touchIdReason: NSLocalizedString("SCREEN_LOCK_REASON_UNLOCK_SCREEN_LOCK_TOUCH_ID",
+                                                                        comment: "Description of how and why Signal iOS uses Touch ID to unlock 'screen lock'."),
+                                       faceIdReason: NSLocalizedString("SCREEN_LOCK_REASON_UNLOCK_SCREEN_LOCK_FACE_ID",
+                                                                       comment: "Description of how and why Signal iOS uses Face ID to unlock 'screen lock'."),
+
+                                       completion: { (outcome: OWSScreenLockOutcome) in
+                                        AssertIsOnMainThread()
+
+                                        switch outcome {
+                                        case .failure(let error):
+                                            failure(self.authenticationError(errorDescription: error))
+                                        case .success:
+                                            success()
+                                        case .cancel:
+                                            cancel()
+                                        }
+        })
+    }
+
+    // On failure, completion is called with an error argument.
+    // On success or cancel, completion is called with nil argument.
+    // Success and cancel can be differentiated by consulting
+    // isScreenLockEnabled.
+    private func tryToVerifyLocalAuthentication(defaultReason: String,
+                                                     touchIdReason: String,
+                                                     faceIdReason: String,
+                                                     completion completionParam: @escaping ((OWSScreenLockOutcome) -> Void)) {
+
+        // Ensure completion is always called on the main thread.
+        let completion = { (outcome: OWSScreenLockOutcome) in
+            switch outcome {
+            case .failure(let error):
+                Logger.error("\(self.TAG) local authentication failed with error: \(error)")
+            default:
+                break
+            }
+            DispatchQueue.main.async {
+                completionParam(outcome)
+            }
+        }
+
+        let context = screenLockContext()
+        let defaultErrorDescription = NSLocalizedString("SCREEN_LOCK_ENABLE_UNKNOWN_ERROR",
+                                                        comment: "Indicates that an unknown error occurred while using Touch ID or Face ID.")
+
+        var authError: NSError?
+        let canEvaluatePolicy = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &authError)
```

Comment:
First there's `canEvaluatePolicy` and then there's `evaluatePolicy`.  To my surprise, `canEvaluatePolicy` uses mostly errors to communicate, not `canEvaluatePolicy`.  Still, I think I'm going to rework this a bit in the next PR.